validation: Send correct notification during snapshot completion

mzumsande opened the pull request:
After AssumeUtxo background sync is completed in a `ActivateBestChain()` call, the `GetRole()` function called with `BlockConnected()` returns `ChainstateRole::NORMAL` instead of `ChainstateRole::BACKGROUND` for this chainstate.
This would make the wallet (which ignores `BlockConnected` notifications for the background chainstate) process it, change `m_last_block_processed_height` to the (ancient) snapshot height, and display an incorrect balance.

Fix this by caching the chainstate role before calling `ActivateBestChainStep()`.
Also contains a test for this situation that fails on master.

Fixes #31546

mzumsande committed:
validation: Send correct notification during snapshot completion

If AssumeUtxo background sync is completed in this
ActivateBestChain() call, the GetRole() function
returns "normal" instead of "background" for this chainstate.
This would make the wallet (which ignores BlockConnected
notifcation for the background chainstate) process it, change
m_last_block_processed_height, and display an incorrect
balance.

mzumsande committed:
test: add functional test for balance after snapshot completion

Use a third node for this, which doesn't get restarted like the second
node.
This test would fail without the previous commit.

fjahr commented on src/validation.cpp:
```diff
@@ -3526,6 +3526,10 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
+                // BlockConnected signals must be sent for the original role;
+                // in case snapshot validation is completed during ActivateBestChainStep, the
+                // result of GetRole() changes from BACKGROUND to NORMAL.
+               ChainstateRole chainstate_role{this->GetRole()};
```
nit: could have been `const`

mzumsande replied:
done

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -57,6 +59,8 @@ def run_test(self):
         n0 = self.nodes[0]
         n1 = self.nodes[1]
+        n2 = self.nodes[2]
+
```
nit: newline not necessary imo

mzumsande replied:
removed, that was not on purpose

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -192,6 +198,18 @@ def run_test(self):
+        self.wait_until(lambda: len(n2.getchainstates()['chainstates']) == 1)
+
+        self.log.info("Check balance at third node")
+        ensure_for(duration=1, f=lambda: (n2.getbalance() == 34))
```
I like to see `ensure_for` getting used but I think I would have opted for `self.wait_until(lambda: (n2.getbalance() == 34), timeout=5)` or so here. Is there a case I am missing where the balance might change after having been 34 that you are covering with this?

mzumsande replied:
I used `ensure_for` because otherwise the added test might not always have failed on master before (if the first check had been done immediately, before the incorrect notification was processed by the wallet, the balance would have still been 34).

fjahr reviewed (commented):
tACK ca0753160a8ebc113e08d62c7b6cbe8fa98455e6

Good catch, feel free to ignore my comments unless you have to retouch. I confirmed that the test fails on master and reviewed the code and surrounding behavior.

mzumsande commented:
[ca07531](https://github.com/bitcoin/bitcoin/commit/ca0753160a8ebc113e08d62c7b6cbe8fa98455e6) to [bc43eca](https://github.com/bitcoin/bitcoin/commit/bc43ecaf6dc0830a27296d3a29428814fed07bb1):
Addressed feedback by @fjahr

fjahr commented:
re-ACK bc43ecaf6dc0830a27296d3a29428814fed07bb1

furszy reviewed (commented):
Code review ACK bc43ecaf6dc

TheCharlatan commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -57,6 +59,7 @@ def run_test(self):
         """
```
Should the docs be updated to mention the added node?

mzumsande replied:
missed this, it could probably be done in a follow-up such as #30455

sedited reviewed (approved):
lgtm ACK bc43ecaf6dc0830a27296d3a29428814fed07bb1

achow101 commented:
ACK bc43ecaf6dc0830a27296d3a29428814fed07bb1
